be/src/runtime/tablets_channel.cpp

```diff
@@ -49,115 +58,168 @@ TabletsChannel::~TabletsChannel() {
-    std::vector<uint32_t> row_indexes(chunk.num_rows());
-    std::vector<uint32_t> channel_row_idx_start_points(channel_size + 1);
-    {
-        // compute row indexes for each channel
-        channel_row_idx_start_points.assign(channel_size + 1, 0);
-        for (uint32_t i = 0; i < params.tablet_ids_size(); ++i) {
-            uint32_t channel_index = _tablet_id_to_sorted_indexes[params.tablet_ids(i)];
-            channel_row_idx_start_points[channel_index]++;
-        }
+    auto res = _create_write_context(request, response, done);
+    if (!res.ok()) {
+        res.status().to_protobuf(response->mutable_status());
+        return;
+    } else {
+        response->mutable_status()->set_status_code(TStatusCode::OK);
+    }
 
-        // NOTE: we make the last item equal with number of rows of this chunk
-        for (int i = 1; i <= channel_size; ++i) {
-            channel_row_idx_start_points[i] += channel_row_idx_start_points[i - 1];
-        }
+    auto context = std::move(res).value();
+    auto channel_size = _tablet_id_to_sorted_indexes.size();
+    auto tablet_ids = request.tablet_ids().data();
+    auto tablet_ids_size = request.tablet_ids_size();
+    auto channel_row_idx_start_points = context->_channel_row_idx_start_points.get();
+    auto row_indexes = context->_row_indexes.get();
+    auto& chunk = context->_chunk;
+    DCHECK_EQ(channel_size, _delta_writers.size());
 
-        for (int i = params.tablet_ids_size() - 1; i >= 0; --i) {
-            uint32_t channel_index = _tablet_id_to_sorted_indexes[params.tablet_ids(i)];
-            row_indexes[channel_row_idx_start_points[channel_index] - 1] = i;
-            channel_row_idx_start_points[channel_index]--;
-        }
-    }
+    auto count_down_latch = BThreadCountDownLatch(1);
+
+    context->set_count_down_latch(&count_down_latch);
 
     for (int i = 0; i < channel_size; ++i) {
         size_t from = channel_row_idx_start_points[i];
         size_t size = channel_row_idx_start_points[i + 1] - from;
         if (size == 0) {
-            // no data for this channel continue;
             continue;
         }
-        auto tablet_id = params.tablet_ids(row_indexes[from]);
+        auto tablet_id = tablet_ids[row_indexes[from]];
         auto it = _delta_writers.find(tablet_id);
-        if (it == std::end(_delta_writers)) {
-            return Status::InternalError(strings::Substitute("unknown tablet to append data, tablet=$0", tablet_id));
-        }
-        {
-            std::lock_guard<std::mutex> l(_tablet_locks[tablet_id & k_shard_size]);
-            auto st = it->second->write(chunk, row_indexes.data(), from, size);
-            if (!st.ok()) {
-                it->second->abort();
-                return st;
+        DCHECK(it != _delta_writers.end());
+        auto& delta_writer = it->second;
+
+        AsyncDeltaWriterRequest req;
+        req.chunk = &context->_chunk;
+        req.indexes = row_indexes + from;
```

Comment:
What's the meaning of the row_indexes? Does it about the row to pick from request to delta_writer?

Reply:
It's inherited from the original code, I don't know exactly what it means